disvae/losses.py

```diff
@@ -229,3 +325,32 @@ def _kl_normal_loss(mean, logvar, storer=None):
         for i in range(latent_dim):
             storer['kl_loss_' + str(i)] += latent_kl[i].item()
     return total_kl
+
+
+def _permute_dims(latent_dist):
+    """
+    Implementation of Algorithm 1 in ref [1]. Randomly permutes the sample from
+    q(z) (latent_dist) across the batch for each of the latent dimensions (mean
+    and log_var).
+
+    Parameters
+    ----------
+    latent_dist : torch.Tensor
+        sufficient statistics of the latent dimension. E.g. for gaussian
+        (mean, log_var), each of shape : (batch_size, latent_dim).
+
+    References :
+        [1] Kim, Hyunjik, and Andriy Mnih. "Disentangling by factorising."
+        arXiv preprint arXiv:1802.05983 (2018).
+
+    """
+
+    perm = torch.zeros(latent_dist.size())
+    dim_j, batch_size, dim_z = latent_dist.size()
+
+    for j in range(dim_j):
+        for z in range(dim_z):
+            pi = torch.randperm(batch_size)
```

Comment:
not sure if indices have to be on CUDA, if yes then you are missing `device=device`

Reply:
Not sure what you mean here by `device=device`? Do you mean it should be `pi = torch.randperm(batch_size).to(latent_dist.device)`?

Reply:
yes might be (not sure whether indices have to be on CUDA, but I guess so)
And I was saying `torch.randperm(batch_size, device=latent_dist.device)` because if you use `.to()` you first allocate on CPU then have to move it to GPU